eqcorrscan/utils/correlate.py

```diff
@@ -24,18 +24,290 @@
 from __future__ import print_function
 from __future__ import unicode_literals
 
-import numpy as np
+import contextlib
 import ctypes
-from future.utils import native_str
+from multiprocessing import Pool as ProcessPool, cpu_count
+from multiprocessing.pool import ThreadPool
 
-from multiprocessing import Pool
+import numpy as np
+from future.utils import native_str
 from scipy.fftpack.helper import next_fast_len
+
 from eqcorrscan.utils.libnames import _load_cdll
 
+XCOR_FUNCS = {}  # cache of functions for doing cross correlations
+
+# methods added to each xcorr func registered
+# these implement the stream interface
+XCORR_STREAM_METHODS = ('multithread', 'multiprocess', 'concurrent',
+                        'stream_xcorr')
+
+# these implement the array interface
+XCOR_ARRAY_METHODS = ('array_xcorr')
+
+
+# ------------------ Context manager for switching out default
+class _Context:
+    """ class for permanently or temporarily changing items in a dict """
+
+    def __init__(self, cache, value_to_switch):
+        """
+        :param cache:
+        :param value_to_switch:
+        """
+        self.cache = cache
+        self.value_to_switch = value_to_switch
+        self.previous_value = None
+
+    def __call__(self, new_value, *args, **kwargs):
+        """ # TODO change docs if this ever becomes general use
```

Comment:
How do you see this becoming general use, do you mean if we expose this to the users? Are we exposing it and will they use it?

Reply:
If we want to have something else that can be toggled based on context, maybe number of cores? My note is just to remove the specific mention of normxcorr if we start using it for other things too. Until/if that happens, it is better to leave the doc string the way it as in order to spoof the _Context instance assigned to the name "set_xcorr" into looking like a function